// scanbot-web-sdk/component/model/document-scanner-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentScannerConfiguration = void 0;
class DocumentScannerConfiguration {
    constructor() {
        /**
         * The container HTML element of Scanbot Web SDK camera. Required
         */
        this.containerId = undefined;
        /**
         * The minimum score in percent (0 - 100) of the perspective
         * distortion to accept a detected document. Default is 75.0.
         */
        this.acceptedAngleScore = 75;
        /**
         * The minimum document width or height in percent (0-100)
         * of the screen size to accept a detected document. Default is 80.0.
         */
        this.acceptedSizeScore = 75;
        /**
         * If auto-capture is enabled and when a document is detected, it will be automatically
         * captured when conditions are good and the auto-snapping time-out elapses.
         */
        this.autoCaptureEnabled = true;
        /**
         * Controls the auto-capture speed. Sensitivity must be within the 0..1 range.
         * A value of 1.0 triggers automatic capturing immediately,
         * a value of 0.0 delays the automatic by 3 seconds. The default value is 0.66 (2 seconds)
         */
        this.autoCaptureSensitivity = 0.66;
        /**
         * Determines whether to ignore the OK_BUT_BAD_ASPECT_RATIO detection status.
         * By default BadAspectRatio is not ignored.
         */
        this.ignoreBadAspectRatio = true;
        this.text = {
            hint: {
                OK: "Don't move. Capturing document...",
                OK_SmallSize: 'Move closer',
                OK_BadAngles: 'Turn your device to<br>have a more rectangular outline',
                OK_BadAspectRatio: 'Wrong aspect ratio - rotate your device',
                OK_OffCenter: 'Move to the center',
                Error_NothingDetected: 'Searching for document...',
                Error_Brightness: 'Poor light<br>Searching for document...',
                Error_Noise: 'Noisy background<br>Searching for document...',
            }
        };
        this.style = {
            /**
             * Optional style outline parameters of the Document Outline (polygon + label) widget
             */
            outline: {
                polygon: {
                    stroke: undefined,
                    strokeWidth: undefined,
                    fill: undefined
                },
                label: {
                    position: undefined,
                    top: undefined,
                    left: undefined,
                    transform: undefined,
                    textAlign: undefined,
                    backgroundColor: undefined,
                    color: undefined,
                    borderRadius: undefined,
                    padding: undefined,
                    fontFamily: undefined,
                    fontSize: undefined
                }
            },
            captureButton: {
                color: "#FFF"
            }
        };
        /**
         * ScanbotCameraView video constraints
         */
        this.videoConstraints = {
            width: {
                ideal: DocumentScannerConfiguration.DEFAULT_VIDEO_WIDTH
            },
            height: {
                ideal: DocumentScannerConfiguration.DEFAULT_VIDEO_HEIGHT
            }
        };
        /**
         * Error callback of the SDK. Should be overriden if you wish custom error handling. Default is empty
         */
        this.onError = function (e) { };
        /**
         * Document detected callback. Continuous callback,
         * stopped only when detection is stopped or camera widget is disposed.
         * Result contains the following properties:
         * detectionStatus: "OK",
         * success: true,
         * cropped: "data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD...",
         * original: "data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD...",
         * polygon: Array(4)}
         */
        this.onDocumentDetected = function (e) { };
    }
    static fromJson(json) {
        const result = new DocumentScannerConfiguration();
        const resultAny = result;
        Object.keys(json).forEach(k => { resultAny[k] = json[k]; });
        result.videoConstraints.facingMode = 'environment';
        result.videoConstraints.resizeMode = 'none';
        return result;
    }
}
exports.DocumentScannerConfiguration = DocumentScannerConfiguration;
DocumentScannerConfiguration.DEFAULT_VIDEO_WIDTH = 3840;
DocumentScannerConfiguration.DEFAULT_VIDEO_HEIGHT = 2160;
