// scanbot-web-sdk/component/view/view-finder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewFinder = void 0;
const compat_1 = __importDefault(require("preact/compat"));
const preact_1 = require("preact");
const utils_1 = require("../utils/utils");
class ViewFinder extends compat_1.default.Component {
    constructor() {
        super();
    }
    render() {
        return preact_1.h("div", { style: this.props.style.overlay, class: "scanbot-camera-overlay" },
            preact_1.h("div", { style: this.props.style.viewfinder, class: "scanbot-camera-viewfinder" }));
    }
    static create(width, height) {
        const finderWidth = 100;
        const finderHeight = 100;
        return preact_1.h(ViewFinder, { style: {
                overlay: {
                    backgroundColor: "black",
                    opacity: 0.7,
                    position: "absolute",
                    width: utils_1.Utils.toPixelSize(width),
                    height: utils_1.Utils.toPixelSize(height),
                    mixBlendMode: "hard-light"
                },
                viewfinder: {
                    width: utils_1.Utils.toPixelSize(finderWidth),
                    height: utils_1.Utils.toPixelSize(finderHeight),
                    backgroundColor: "gray",
                    borderRadius: "10px",
                    border: "2px solid white",
                    marginLeft: utils_1.Utils.toPixelSize(width / 2 - finderWidth / 2),
                    marginTop: utils_1.Utils.toPixelSize(height / 2 - finderHeight / 2),
                }
            } });
    }
}
exports.ViewFinder = ViewFinder;
